components/brave_wallet/browser/asset_discovery_manager.cc

```diff
@@ -330,6 +337,274 @@ void AssetDiscoveryManager::MergeDiscoveredEthAssets(
+void AssetDiscoveryManager::FetchNFTsFromSimpleHash(
+    const std::string& account_address,
+    const std::vector<std::string>& chain_ids,
+    mojom::CoinType coin,
+    FetchNFTsFromSimpleHashCallback callback) {
+  if (!(coin == mojom::CoinType::ETH || coin == mojom::CoinType::SOL)) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  GURL url = GetSimpleHashNftsByWalletUrl(account_address, chain_ids);
+  if (!url.is_valid()) {
+    std::move(callback).Run({});
+    return;
+  }
+  std::vector<mojom::BlockchainTokenPtr> nfts_so_far = {};
+  auto internal_callback =
+      base::BindOnce(&AssetDiscoveryManager::OnFetchNFTsFromSimpleHash,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(nfts_so_far),
+                     coin, std::move(callback));
+
+  api_request_helper_->Request("GET", url, "", "", true,
+                               std::move(internal_callback));
+}
+
+void AssetDiscoveryManager::OnFetchNFTsFromSimpleHash(
+    std::vector<mojom::BlockchainTokenPtr> nfts_so_far,
+    mojom::CoinType coin,
+    FetchNFTsFromSimpleHashCallback callback,
+    APIRequestResult api_request_result) {
+  if (!api_request_result.Is2XXResponseCode()) {
+    std::move(callback).Run(std::move(nfts_so_far));
+    return;
+  }
+
+  // Invalid JSON becomes an empty string after sanitization
+  if (api_request_result.body().empty()) {
+    std::move(callback).Run(std::move(nfts_so_far));
+    return;
+  }
+
+  absl::optional<std::pair<GURL, std::vector<mojom::BlockchainTokenPtr>>>
+      result = ParseNFTsFromSimpleHash(api_request_result.value_body(), coin);
+  if (!result) {
+    std::move(callback).Run(std::move(nfts_so_far));
+    return;
+  }
+
+  for (auto& token : result.value().second) {
+    nfts_so_far.push_back(std::move(token));
+  }
+
+  // If there is a next page, fetch it
+  if (result.value().first.is_valid()) {
+    auto internal_callback =
+        base::BindOnce(&AssetDiscoveryManager::OnFetchNFTsFromSimpleHash,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(nfts_so_far),
+                       coin, std::move(callback));
+    api_request_helper_->Request("GET", result.value().first, "", "", true,
+                                 std::move(internal_callback));
+    return;
+  }
+
+  // Otherwise, return the nfts_so_far
+  std::move(callback).Run(std::move(nfts_so_far));
+}
+
+absl::optional<std::pair<GURL, std::vector<mojom::BlockchainTokenPtr>>>
+AssetDiscoveryManager::ParseNFTsFromSimpleHash(const base::Value& json_value,
+                                               mojom::CoinType coin) {
+  // Only ETH and SOL NFTs are supported.
+  if (!(coin == mojom::CoinType::ETH || coin == mojom::CoinType::SOL)) {
+    return absl::nullopt;
+  }
+
+  const base::Value::Dict* dict = json_value.GetIfDict();
+  if (!dict) {
+    return absl::nullopt;
+  }
+
+  GURL nextURL;
+  auto* next = dict->FindString("next");
```

Comment:
Is this just a pagination URL effectively? Simplehash's docs don't define very well what this actually is for.

Reply:
That's right